src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc

```diff
@@ -95,19 +96,23 @@ BootstrapRequest<I>::BootstrapRequest(librados::IoCtx &local_io_ctx,
                                       const std::string &remote_mirror_uuid,
                                       Journaler *journaler,
                                       MirrorPeerClientMeta *client_meta,
-                                      Context *on_finish)
+                                      Context *on_finish,
+				      rbd::mirror::ProgressContext *progress_ctx)
   : m_local_io_ctx(local_io_ctx), m_remote_io_ctx(remote_io_ctx),
     m_local_image_ctx(local_image_ctx), m_local_image_name(local_image_name),
     m_remote_image_id(remote_image_id), m_global_image_id(global_image_id),
     m_work_queue(work_queue), m_timer(timer), m_timer_lock(timer_lock),
     m_local_mirror_uuid(local_mirror_uuid),
     m_remote_mirror_uuid(remote_mirror_uuid), m_journaler(journaler),
-    m_client_meta(client_meta), m_on_finish(on_finish) {
+    m_client_meta(client_meta), m_on_finish(on_finish),
+    m_progress_ctx(progress_ctx) {
 }
 
 template <typename I>
 BootstrapRequest<I>::~BootstrapRequest() {
   assert(m_remote_image_ctx == nullptr);
+
+  delete(m_progress_ctx);
```

Comment:
I was thinking ```ImageReplayer``` would have a ```ImageReplayerProgressContext``` proxy which it passes to ```BootstrapRequest``` as a ```ProgressContext*``` (owned by ImageReplayer -- could even be a member variable of ImageReplayer since the lifetime is intertwined).